### PR TITLE
Check min version of numpy, scipy, and librosa in common.py

### PR DESCRIPTION
**Pull request by ssnl**

### CODE REVIEW

File: test/common.py [hunk 1]
```diff
@@ -68,27 +69,26 @@ def _check_module_exists(name):
...
+            return False
     else:  # Python >= 3.4
         import importlib
-        spec = importlib.util.find_spec(name)
-        return spec is not None
+        if importlib.util.find_spec(name) is None:
+            return False
+    if min_version is None:
+        return True
+    module_version = parse_version(eval("__import__({}).__version__").format(name))
```
Comment (zou3519) [inline test/common.py]: do we need `eval`? can't you try to import it and then get the .__version__ in code?

File: test/common.py [hunk 2]
```diff
@@ -68,27 +69,26 @@ def _check_module_exists(name):
...
+            return False
     else:  # Python >= 3.4
         import importlib
-        spec = importlib.util.find_spec(name)
-        return spec is not None
+        if importlib.util.find_spec(name) is None:
+            return False
+    if min_version is None:
+        return True
+    module_version = parse_version(eval("__import__('{}').__version__").format(name))
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/common.py]: :( :(

File: test/common.py [hunk 3]
```diff
@@ -57,37 +58,29 @@ def run_tests(argv=UNITTEST_ARGS):
...
 
-def _check_module_exists(name):
-    r"""Returns if a top-level module with :attr:`name` exists *without**
-    importing it. This is generally safer than try-catch block around a
-    `import X`. It avoids third party libraries breaking assumptions of some of
-    our tests, e.g., setting multiprocessing start method when imported
-    (see librosa/#747, torchvision/#544).
+def _check_module_exists(name, min_version=None):
+    r"""Returns if a top-level module with :attr:`name` exists and satisfy a
+    minium version of :attr:`min_version` (if set). Ideally, this should not do
```
Comment (zou3519) [inline test/common.py]: s/minium/minimum

File: test/common.py [hunk 4]
```diff
@@ -57,37 +58,29 @@ def run_tests(argv=UNITTEST_ARGS):
...
+        m = __import__(name)
+    except ImportError:
+        return False
+    if min_version is None:
+        return True
+    return parse_version(m.__version__) >= parse_version(min_version)
 
-# On Py2, importing librosa 0.6.1 triggers a TypeError (if using newest joblib)
 # see librosa/librosa#729.
-# TODO: allow Py2 when librosa 0.6.2 releases
```
Comment (zou3519) [inline test/common.py]: Is librosa 0.6.2 not going to be released? Maybe keep a todo?

File: test/common.py [hunk 5]
```diff
@@ -57,37 +58,29 @@ def run_tests(argv=UNITTEST_ARGS):
...
+    try:
+        m = __import__(name)
+    except ImportError:
+        return False
+    if min_version is None:
+        return True
+    return parse_version(m.__version__) >= parse_version(min_version)
 
-# On Py2, importing librosa 0.6.1 triggers a TypeError (if using newest joblib)
 # see librosa/librosa#729.
```
Comment (zou3519) [inline test/common.py]: This is out of order now

File: test/common.py [hunk 6]
```diff
@@ -57,37 +58,29 @@ def run_tests(argv=UNITTEST_ARGS):
...
-    `import X`. It avoids third party libraries breaking assumptions of some of
-    our tests, e.g., setting multiprocessing start method when imported
-    (see librosa/#747, torchvision/#544).
+def _check_module_exists(name, min_version=None):
+    r"""Returns if a top-level module with :attr:`name` exists and satisfy a
+    minium version of :attr:`min_version` (if set). Ideally, this should not do
+    an actual import to avoid third party libraries breaking assumptions of some
+    of our tests, e.g., setting multiprocessing start method when imported
+    (see librosa/#747, torchvision/#544). However, we guard our multiprocessing
+    tests with `mp = mp.get_context(...)` so this shouldn't be too big a
```
Comment (zou3519) [inline test/common.py]: I'm not too sure about the consequences of this. Does `pkg_resources.get_distribution("numpy").version` work? (found this on stackoverflow)

File: test/common.py [hunk 7]
```diff
@@ -58,38 +60,42 @@ def run_tests(argv=UNITTEST_ARGS):
 IS_PPC = platform.machine() == "ppc64le"
 
 
-def _check_module_exists(name):
-    r"""Returns if a top-level module with :attr:`name` exists *without**
-    importing it. This is generally safer than try-catch block around a
-    `import X`. It avoids third party libraries breaking assumptions of some of
-    our tests, e.g., setting multiprocessing start method when imported
+def _check_module_exists(name, min_version=None):
+    r"""Returns if a top-level module with :attr:`name` exists and satisfy a
```
Comment (zou3519) [inline test/common.py]: nit: s/satisfy/satisfies

**Review by zou3519 — APPROVED** [review]
lgtm!

**Review by weiyangfb — APPROVED** [review]
lgtm

File: test/test_nn.py [hunk 8]
```diff
@@ -5263,8 +5263,7 @@ def test_affine_grid(self):
             self.assertEqual(out_cpu, out_cuda)
             self.assertEqual(input_cpu.grad, input_gpu.grad)
 
-    @unittest.skipIf((not TEST_NUMPY) or (not TEST_SCIPY) or (scipy.__version__ < '1.0.0'),
-                     "Scipy v1.0 and/or numpy not found")
+    @unittest.skipIf((not TEST_NUMPY) or (not TEST_SCIPY), "Scipy and/or numpy not found")
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_nn.py]: What do you mean? a decorator should modify the function's behavior at runtime

[the remark this answers is not in the document]
Comment (weiyangfb) [reply test/test_nn.py]: oh, I see what you meant. Yeah, I agree decorator will look much nicer here and no compromise on efficiency. I meant a decorator like [this](https://github.com/pytorch/pytorch/blob/b14f2e899c92f541644e393b2513aa60aeb9196f/test/common.py#L111-L120) that run the function first and error out will be slow

[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_nn.py]: Oh, I didn't know that's what that did... a general rule of thumb I have is that decorators shouldn't run the function, they should modify the behavior of the function when it is run

[the remark this answers is not in the document]
Comment (SsnL) [reply test/test_nn.py]: I am 99.9999% sure that `skipIf` isn't implemented like `skipIfNoLapack`. It probably looks like:
```py
def skipIf(cond, msg):
  if cond:
    return lambda fn: fn
  def skipped(*args, **kwargs):
    raise Skip(msg)
  return lambda fn: functools.wraps(fn)(skipped)
```

**Pull request closed without merging**